compiler/plugins/target/AMD-AIE/aie/AMDAIECreatePathFindFlows.cpp

```diff
@@ -13,671 +14,566 @@
-  d.walk([&](ConnectOp connect) {
-    if (auto sw = connect->getParentOfType<SwitchboxOp>()) {
-      // Constraint: memtile stream switch constraints
-      if (auto tile = sw.getTileOp();
-          tile.isMemTile() &&
-          !deviceModel.isLegalMemtileConnection(
-              tile.getCol(), tile.getRow(), toStrmT(connect.getSourceBundle()),
-              connect.getSourceChannel(), toStrmT(connect.getDestBundle()),
-              connect.getDestChannel())) {
-        problemConnects.push_back(connect);
-      }
+      getDeviceModel(static_cast<AMDAIEDevice>(device.getDevice()));
+  for (auto switchbox : make_early_inc_range(device.getOps<SwitchboxOp>())) {
+    auto retVal = switchbox->getOperand(0);
+    auto tileOp = retVal.getDefiningOp<TileOp>();
+
+    if (!deviceModel.isShimNOCTile(tileOp.getCol(), tileOp.getRow())) continue;
+    // Check if the switchbox is empty
+    if (&switchbox.getBody()->front() == switchbox.getBody()->getTerminator())
+      continue;
+
+    ShimMuxOp shimMuxOp = nullptr;
+    for (auto shimmux : device.getOps<ShimMuxOp>()) {
+      if (shimmux.getTile() != tileOp) continue;
+      shimMuxOp = shimmux;
+      break;
+    }
+    if (!shimMuxOp) {
+      builder.setInsertionPointAfter(tileOp);
+      shimMuxOp = getOrCreateShimMux(builder, device, tileOp.getCol());
     }
-  });
 
-  for (auto connect : problemConnects) {
-    auto swBox = connect->getParentOfType<SwitchboxOp>();
-    OpBuilder::InsertionGuard g(builder);
-    builder.setInsertionPoint(connect);
-    auto northSw = getSwitchbox(d, swBox.colIndex(), swBox.rowIndex() + 1);
-    if (auto southSw = getSwitchbox(d, swBox.colIndex(), swBox.rowIndex() - 1);
-        !attemptFixupMemTileRouting(builder, northSw, southSw, connect)) {
-      return signalPassFailure();
+    for (Operation &op : switchbox.getConnections().getOps()) {
+      // check if there is MM2S DMA in the switchbox of the 0th row
+      if (auto pktrules = llvm::dyn_cast<PacketRulesOp>(op);
+          pktrules &&
+          toStrmT(pktrules.getSourceBundle()) == StrmSwPortType::DMA) {
+        // If there is, then it should be put into the corresponding shimmux
+        OpBuilder::InsertionGuard g(builder);
+        Block &b0 = shimMuxOp.getConnections().front();
+        builder.setInsertionPointToStart(&b0);
+        pktrules.setSourceBundle(toWireB(StrmSwPortType::SOUTH));
+        if (pktrules.getSourceChannel() == 0) {
+          pktrules.setSourceChannel(3);
+          builder.create<ConnectOp>(builder.getUnknownLoc(),
+                                    toWireB(StrmSwPortType::DMA), 0,
+                                    toWireB(StrmSwPortType::NORTH), 3);
+        } else if (pktrules.getSourceChannel() == 1) {
+          pktrules.setSourceChannel(7);
+          builder.create<ConnectOp>(builder.getUnknownLoc(),
+                                    toWireB(StrmSwPortType::DMA), 1,
+                                    toWireB(StrmSwPortType::NORTH), 7);
+        }
```

Comment:
What's this doing?

Reply:
![image](https://github.com/user-attachments/assets/14d4e893-69c8-42e4-a102-f27a0ae6448a)